src/tree/identity_tree.rs

```diff
@@ -0,0 +1,610 @@
+                .or_else(|| {
+                    let (depth, offset) =
+                        storage_idx_to_coords(sibling_idx as usize);
+                    Some(self.tree.get_node(depth, offset))
+                })
+                .expect("Could not find node in tree");
+
+            proof.push(if node_idx % 2 == 0 {
+                Branch::Right(sibling)
+            } else {
+                Branch::Left(sibling)
+            });
+
+            node_idx = (node_idx - 1) / 2;
+        }
+
+        Ok(semaphore::merkle_tree::Proof(proof))
+    }
+
+    pub fn insert(&mut self, index: u32, leaf: Hash) -> eyre::Result<()> {
+        // Check if the leaf laready exists
+        if self.leaves.contains_key(&leaf) {
+            return Err(eyre!("Leaf already exists"));
+        }
+
+        self.leaves.insert(leaf, index);
+
+        // We can expect here because the `reallocate` implementation for Vec<H::Hash> as DynamicTreeStorage does not fail
+        self.tree.push(leaf).expect("Failed to insert into tree");
+
+        Ok(())
+    }
+
+    pub fn remove(&mut self, index: usize) {
+        let leaf = self.tree.get_leaf(index);
+        self.leaves.remove(&leaf);
+        self.tree.set_leaf(index, Hash::ZERO);
+    }
+
+    // Appends new leaf updates and newly calculated intermediate nodes to the tree updates
+    pub fn append_updates(&mut self, root: Root, leaf_updates: LeafUpdates) {
+        // Update leaves
+        match &leaf_updates {
+            LeafUpdates::Insert(updates) => {
+                for (idx, val) in updates.iter() {
+                    self.leaves.insert(*val, idx.into());
+                }
+            }
+            LeafUpdates::Delete(updates) => {
+                for (_, val) in updates.iter() {
+                    self.leaves.remove(val);
+                }
+            }
+        }
+
+        let mut updates = HashMap::new();
+        let mut node_queue = VecDeque::new();
+
+        // Convert leaf indices into storage indices and insert into updates
+        let leaves: Leaves = leaf_updates.into();
+        for (leaf_idx, hash) in leaves.into_iter() {
+            let storage_idx = leaf_to_storage_idx(*leaf_idx, self.tree.depth());
+            updates.insert(storage_idx.into(), hash);
+
+            // Queue the parent index
+            let parent_idx = (storage_idx - 1) / 2;
+            node_queue.push_front(parent_idx);
+        }
+
+        let prev_update = if let Some(update) = self.tree_updates.iter().last()
+        {
+            //TODO: Use a more efficient approach than to clone the last update
+            update.1.clone()
+        } else {
+            HashMap::new()
+        };
+
+        while let Some(node_idx) = node_queue.pop_back() {
+            // Check if the parent is already in the updates hashmap, indicating it has already been calculated
+            //TODO: note why we set to 0 if idx is 0
+            let parent_idx = if node_idx == 0 { 0 } else { (node_idx - 1) / 2 };
```

Comment:
Replacing `0` with `continue` here and skipping that one iteration seems to work here (tests in this file pass)